packages/langium/src/documents/document-builder.ts

```diff
@@ -54,55 +50,72 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
+    protected async validate(document: LangiumDocument, cancelToken = CancellationToken.None, forceDiagnostics = false): Promise<Diagnostic[]> {
+        let diagnostics: Diagnostic[] = [];
         const validator = this.documentValidator;
-        if (this.connection) {
-            diagnostics = validator.validateDocument(document);
-            // Send the computed diagnostics to VS Code.
-            this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+        if (this.connection || forceDiagnostics) {
+            diagnostics = await validator.validateDocument(document, cancelToken);
+            if (this.connection && !cancelToken.isCancellationRequested) {
+                // Send the computed diagnostics to VS Code.
+                this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+            }
+        }
+        if (!cancelToken.isCancellationRequested) {
+            document.state = DocumentState.Validated;
         }
         return diagnostics;
     }
 
-    documentChanged(uri: URI): void {
+    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
         this.langiumDocuments.invalidateDocument(uri);
         const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
-        this.indexManager.update(newDocument);
-        this.langiumDocuments.all.filter(doc => this.isAffected(doc, uri)).forEach(
-            doc => {
-                this.build(doc);
-                this.indexManager.update(doc);
+        await delayNextTickAfter();
+        if (cancelToken.isCancellationRequested) {
+            return;
+        }
+        await this.buildDocument(newDocument, cancelToken);
+    }
+
+    protected async buildDocument(document: LangiumDocument, cancelToken: CancellationToken): Promise<void> {
+        const affectedDocuments = this.indexManager.getEffectedDocuments(document);
+        affectedDocuments.forEach(e => e.reset());
+        const relevantDocuments = Array.from(new Set([
+            document,
+            ...affectedDocuments,
+            // Also include all documents haven't completed the document lifecycle yet
+            ...this.langiumDocuments.all.filter(e => e.state < DocumentState.Validated)
+        ]));
+        await this.runCancelable(relevantDocuments, DocumentState.Processed, cancelToken, doc => this.process(doc));
+        await this.runCancelable(relevantDocuments, DocumentState.Indexed, cancelToken, doc => this.indexManager.update(doc, cancelToken));
+        await this.runCancelable(relevantDocuments, DocumentState.Validated, cancelToken, doc => this.validate(doc, cancelToken));
+    }
+
+    protected async runCancelable(documents: LangiumDocument[], state: DocumentState, cancelToken: CancellationToken, callback: (document: LangiumDocument) => MaybePromise<unknown>): Promise<void> {
+        if (cancelToken.isCancellationRequested) {
+            return;
+        }
```

Comment:
I think checking the token is not necessary here since it's also done two lines below.